### ISSUE TITLE
nobody2018 - When PartyA can be liquidated, LiquidationFacet.liquidatePendingPositionsPartyA should be called before LiquidationFacet.setSymbolsPrice

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-06-symmetrical-judging#205]
nobody2018

medium

# When PartyA can be liquidated, LiquidationFacet.liquidatePendingPositionsPartyA should be called before LiquidationFacet.setSymbolsPrice

## Summary

When PartyA1 can be liquidated, the process is as follows:

1.  The liquidator sets `liquidationStatus[PartyA1]` to true via `LiquidationFacet.liquidatePartyA`.
2.  The liquidator sets the `liquidationDetails[PartyA1]` via `LiquidationFacet.setSymbolsPrice`.
3.  The liquidator processes [[quoteIds](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacet.sol#L43)](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacet.sol#L43) via `LiquidationFacet.liquidatePositionsPartyA`, and every quote in `quoteIds` array must be created by PartyA1.
4.  The liquidator removes all quotes from the `partyAPendingQuotes[PartyA1]` array via `LiquidationFacet.liquidatePendingPositionsPartyA`.

The order of steps 1, 2, and 3 cannot be changed. But step 4 only needs to be after step 1. There are two issues here:

1.  According to the product design, when an unopened quote is removed, the trading fee should be returned to PartyA. `LiquidationFacetImpl.liquidatePendingPositionsPartyA` lacks the return of trading fee to PartyA. This is unreasonable.
2.  The returned trading fee is added to [[allocatedBalances[partyA1]](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibQuote.sol#L138)](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibQuote.sol#L138). So step 4 should be between steps 1 and 2. This is more fair to PartyB.

## Vulnerability Detail

The code snippet of `LiquidationFacetImpl.liquidatePendingPositionsPartyA` is as follows:

```solidity
File: symmio-core\contracts\facets\liquidation\LiquidationFacetImpl.sol
105:         for (uint256 index = 0; index < quoteLayout.partyAPendingQuotes[partyA].length; index++) {
106:             Quote storage quote = quoteLayout.quotes[
107:                 quoteLayout.partyAPendingQuotes[partyA][index]
108:             ];
109:             if (
110:                 (quote.quoteStatus == QuoteStatus.LOCKED ||
111:                     quote.quoteStatus == QuoteStatus.CANCEL_PENDING) &&
112:                 quoteLayout.partyBPendingQuotes[quote.partyB][partyA].length > 0
113:             ) {
114:                 delete quoteLayout.partyBPendingQuotes[quote.partyB][partyA];
115:                 AccountStorage
116:                 .layout()
117:                 .partyBPendingLockedBalances[quote.partyB][partyA].makeZero();
118:             }
119:             
120:             quote.quoteStatus = QuoteStatus.LIQUIDATED;
121:             quote.modifyTimestamp = block.timestamp;
122:         }
```

Obviously, the above code does not return the trading fee to PartyA. `LiquidationFacetImpl.setSymbolsPrice` uses [[LibAccount.partyAAvailableBalanceForLiquidation](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol#L52-L55)](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol#L52-L55) to get `availableBalance`, which is related to [[allocatedBalances[partyA1]](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibAccount.sol#L83-L84)](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibAccount.sol#L83-L84). Therefore, the returned trading fee should be included in the liquidation.

## Impact

As mentioned in the Summary, when a PartyA is liquidated, the pending quote should be returned to the trading fee which is counted in the liquidation. This is more fair to PartyB.

## Code Snippet

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol#L52-L87

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol#L105-L121

https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibAccount.sol#L78-L86

## Tool used

Manual Review

## Recommendation

Fix for issue 1:

```diff
--- a/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol
+++ b/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol
@@ -108,13 +108,15 @@ library LiquidationFacetImpl {
             ];
             if (
                 (quote.quoteStatus == QuoteStatus.LOCKED ||
-                    quote.quoteStatus == QuoteStatus.CANCEL_PENDING) &&
-                quoteLayout.partyBPendingQuotes[quote.partyB][partyA].length > 0
+                    quote.quoteStatus == QuoteStatus.CANCEL_PENDING)
             ) {
-                delete quoteLayout.partyBPendingQuotes[quote.partyB][partyA];
-                AccountStorage
-                .layout()
-                .partyBPendingLockedBalances[quote.partyB][partyA].makeZero();
+                LibQuote.returnTradingFee(quote.id);
+                if (quoteLayout.partyBPendingQuotes[quote.partyB][partyA].length > 0) {
+                    delete quoteLayout.partyBPendingQuotes[quote.partyB][partyA];
+                    AccountStorage
+                    .layout()
+                    .partyBPendingLockedBalances[quote.partyB][partyA].makeZero();
+                }
             }
             quote.quoteStatus = QuoteStatus.LIQUIDATED;
             quote.modifyTimestamp = block.timestamp;
```

Fix for issue 2:

```diff
--- a/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol
+++ b/symmio-core/contracts/facets/liquidation/LiquidationFacetImpl.sol
@@ -42,6 +42,8 @@ library LiquidationFacetImpl {
                 maLayout.liquidationTimestamp[partyA] + maLayout.liquidationTimeout,
             "LiquidationFacet: Expired signature"
         );
+        if (QuoteStorage.layout().partyAPendingQuotes[partyA].length > 0)
+            revert MustCall_liquidatePositionsPartyA_First();
         for (uint256 index = 0; index < priceSig.symbolIds.length; index++) {
             accountLayout.symbolsPrices[partyA][priceSig.symbolIds[index]] = Price(
                 priceSig.prices[index],
```

Duplicate of #71

**Comment by securitygrid** [comment]
Escalate for 10 usdc
This issue should not be dup of #71. This issue describles that the pending quote should be returned to the trading fee which is counted in the liquidation. step 4 should be between steps 1 and 2. Otherwise, the returned funding fee will not be counted in the liquidation.

**Comment by sherlock-admin2** [comment]
> Escalate for 10 usdc
> This issue should not be dup of #71. This issue describles that the pending quote should be returned to the trading fee which is counted in the liquidation. step 4 should be between steps 1 and 2. Otherwise, the returned funding fee will not be counted in the liquidation.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by panprog** [comment]
I agree that step 4 should be between step 1 and 2, otherwise if it's after step 2, then fee will be added to partyA allocatedBalances, but it is not used anywhere (the balance+upnl is sent to partyB in step 2) and in step 3 it will be reset to 0, so effectively the fee will be lost.

However, I believe that this order of steps (issue 2) is just the implementation details of the fix of the issue 1 (which is a clear dup of #71), because right now there is nothing wrong with the order of liquidation steps (issue 2 is currently non-existant), issue 2 will only appear if the issue 1 is fixed simply by adding fee return, however if it's fixed correctly (the order of steps taken into account), then issue 2 will never appear.

I agree that in most other dups of #71 this fix implementation detail is never mentioned, however this shouldn't make this report a unique finding - it's the same finding, just with better description of the fix.

**Comment by securitygrid** [comment]
I submitted this issue separately because of different impacts and different fixes.

**Comment by panprog** [comment]
> I submitted this issue separately because of different impacts and different fixes.

Do you agree that issue 2 currently doesn't exists?

Let me quote your description of issue 2 from the report:

> 2\. The returned trading fee is added to [[allocatedBalances[partyA1]](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibQuote.sol#L138)](https://github.com/sherlock-audit/2023-06-symmetrical/blob/main/symmio-core/contracts/libraries/LibQuote.sol#L138). So step 4 should be between steps 1 and 2. This is more fair to PartyB.

But right now trading fee **is not** returned during the liquidation of pending quotes, `returnTradingFee` is not called right now, so nothing will change if step 4 is enforced between steps 1 and 2.

Your report implies that the fix for issue 1 is simply to return trading fee in the liquidation of pending quotes, and **only then** the issue 2 will appear. However, the issue 2 will only appear with this exact fix of issue 1, and it only means that such fix for issue 1 is incorrect. The correct fix for issue 1 should take this situation into account, and if fixed correctly, issue 2 will never appear (although it should be considered when implementing fix).

**Comment by securitygrid** [comment]
The reason why #201 and #205 were submitted separately is that the logic when two different entities are liquidated are completely different. Although the two reports  have the same root, both impact and fix are different. 
#201 fixes the problem from the perspective of PartyB, while #205 fixes it from the perspective of PartyA.
No matter how this report will be judged, it will be noticed by the sponsor. That's all right.
Let's leave it to Sherlock:)

**Comment by hrishibhat** [comment]
Result:
Medium
Duplicate of #71

**Comment by sherlock-admin2** [comment]
Escalations have been resolved successfully!

Escalation status:
- [securitygrid](https://github.com/sherlock-audit/2023-06-symmetrical-judging/issues/205/#issuecomment-1651522729): rejected
